sql-plugin/src/main/java/com/nvidia/spark/rapids/JCudfUtil.java

```diff
@@ -0,0 +1,437 @@
+      case STRING:
+        return JCUDF_TYPE_STRING_LENGTH_ESTIMATE;
+      case LIST:
+      case STRUCT:
+        // Return 32 bytes for other types until we decide on how to handle each type.
+      default:
+        throw new IllegalArgumentException(rapidsType + " estimated size is not supported yet.");
+    }
+  }
+
+  /**
+   * Checks if CUDF Kernel can apply optimized row conversion on the Spark schema represented by
+   * {@link CudfUnsafeRow}. Note that the calculation assumes that the {@code cudfUnsafeRow}
+   * is not usable yet (i.e., {@link CudfUnsafeRow#pointTo(long, int)} is not called yet).
+   *
+   * The fixed-width optimized cudf kernel only supports up to 1.5 KB per row.
+   * For fixed-size rows, we know the exact number of bytes used by the columns and the validity
+   * bytes. We can use that value to branch over the size of the output to know which kernel to
+   * call.
+   *
+   * @param cudfUnsafeRow the unsafeRow object that represents the Spark SQL Schema.
+   * @return true if {@link CudfUnsafeRow#getFixedWidthSizeInBytes()} is less than
+   *         {@link #JCUDF_MAX_FIXED_ROW_SIZE_FITS_OPTIMIZED} which is 1528 Bytes.
+   */
+  public static boolean fitsOptimizedConversion(
+      CudfUnsafeRow cudfUnsafeRow) {
+    return !cudfUnsafeRow.isVariableWidthSchema() &&
+        cudfUnsafeRow.getVariableSizeDataOffset() < JCUDF_MAX_FIXED_ROW_SIZE_FITS_OPTIMIZED;
+  }
+
+  /**
+   * @see #fitsOptimizedConversion(CudfUnsafeRow)
+   * In some situations, we can calculate an estimate size for a schema withoput constructing
+   * the object yet. In that case, we can use {@link JCudfRowVisitor} which already processed the
+   * schema.
+   */
+  public static boolean fitsOptimizedConversion(
+      JCudfRowOffsetsEstimator cudfRowOffsetMock) {
+    return !cudfRowOffsetMock.hasVarSizeData() &&
+        cudfRowOffsetMock.getCurrentByteOffset() < JCUDF_MAX_FIXED_ROW_SIZE_FITS_OPTIMIZED;
+  }
+
+  /**
+   * A helper class to calculate the columns, validity, and variable width offsets.
+   * This helper is used to get an estimate size for the row including variable-sized rows.
+   * The implementation does not make any assumptions whether the schema is packed or not.
+   * Unlike {@link JCudfRowVisitor}, this implementation does not cache any calculations.
+   */
+  public static class JCudfRowOffsetsEstimator {
+    Attribute[] attributes;
+    int byteCursor = 0;
+    int varSizeColIndex;
+    int validityBytesOffset = 0;
+
+    private JCudfRowOffsetsEstimator(Attribute[] attrArr) {
+      this.attributes = attrArr;
+      resetMeta();
+    }
+
+    private int calcColOffset(int ind) {
+      int res;
+      Attribute attr = attributes[ind];
+      DataType dataType = attr.dataType();
+      DType rapidsType = GpuColumnVector.getNonNestedRapidsType(dataType);
+      if (isFixedLength(dataType)) {
+        int length = rapidsType.getSizeInBytes();
+        res = alignOffset(byteCursor, length);
```

Comment:
This seems dangerous as it relies on the byteCursor to hold state. On the surface, it seems like you can call:
```
calcColOffset(8);
calcColOffset(12);
calcColOffset(0);
```
But this won't work correctly as it has to be iterated. Should this function have comments saying that it is intended only to be used iteratively or have some indication of that in the name?

Reply:
I noticed this is private, so this might not be as big of a deal as originally thought.

Reply:
To support random access to the fields, we will need to iterate on the columns during the initialization which increases the overhead of the code block.
I added comments to the methods explicitly mentioning that the access has to be done sequentially. This is a tradeoffs between performance/abstraction

Reply:
I like that this is private now and just sets the values once at construction time. It is not the cleanest thing I have seen, but it looks like it works so I am okay with it.